scrapyd_k8s/object_storage/libcloud_driver.py

```diff
@@ -0,0 +1,79 @@
+import os
+import logging
+
+logger = logging.getLogger(__name__)
+
+from libcloud.storage.types import ObjectError, ContainerDoesNotExistError, ObjectDoesNotExistError, InvalidContainerNameError
+from libcloud.storage.providers import get_driver
+
+class LibcloudObjectStorage:
+
+    def __init__(self, config):
+        self._storage_provider = config.joblogs().get('storage_provider')
+        if self._storage_provider is None:
+            logger.error('Storage provider is not defined in the configuration.')
+            raise ValueError('Storage provider is not defined')
+
+        self._container_name = config.joblogs().get('container_name')
+        if self._container_name is None:
+            logger.error('Container name is not set in the configuration.')
+            raise ValueError('Container name is not set')
+
+        args_envs = config.joblogs_provider_args(self._storage_provider)
+        args = {}
+        for arg, env in args_envs.items():
+            env_value = os.getenv(env)
+            if env_value is None:
+                logger.error(f"Environment variable '{env}' for argument '{arg}' is not set.")
+                raise ValueError(f"Environment variable '{env}' for argument '{arg}' is not set.")
+            args[arg] = env_value
```

Comment:
I'm not _quite_ sure what happens here. Is it so that all items listed in the provider args, are expected to be set in the environment? While that would work, I think it would be more usable / easier to use when some items could be set in the configuration file, and some could be gotten from the environment.

A very simple approach would be variable expansion, so `${ABC}` would get the value from the `ABC` environment variable (and the need to allow escaping `$` when needed in the value itself). Or let `region = eu-west-1` set it directly, and `region_env = AWS_REGION` get it from the environment.
_Ideally_ this might be generalizable to the whole configuration file, allowing for more flexibility, but I think that actually would make it too complex (one can still generate the config file if you need it) - but such an approach may still be helpful to think about a generic function that does this with a config section or so.

Reply:
I _would_ be ok with keeping it as it is, provided that we mark this functionality as alpha or so, that users should expect that the configuration approach of this section can still change. Since it would also be nice to converge a bit to getting this merged. (and then work on this later, by you or someone else)

Reply:
I added the alternative, if cloud credentials are not provided as ENV, then it takes values from config, tested for the case you mentioned:
```
    [joblogs.storage.s3]
    key    = ${S3_KEY}
    secret = ${S3_SECRET}
    region = eu-north-1
```
I would discourage any explicit config values to avoid compromising the cloud, but maybe region is not that critical in this case.

Reply:
Super, that sounds great!
There are also other values like endpoint URL, whether to use domain-host-prefixes or not, that may make more sense to put in version control for deployment, vs. secrets. This gives the flexibility for sysadmins to decide that for themselves. :+1: